graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/node/GqlJcrNodeImpl.java

```diff
@@ -638,4 +666,33 @@ private void collectUsages(PropertyIterator references, Collection<GqlUsage> gql
             }
         }
     }
+
+    private String getNodeURL(JCRNodeWrapper node, final String workspace, final Locale locale, boolean findDisplayable)
+            throws RepositoryException {
+        String servlet = "render";
+        String url = Jahia.getContextPath() + "/cms/" + servlet + "/" + workspace + "/" + locale;
+
+        Resource resource = new Resource(node, "html", null, Resource.CONFIGURATION_PAGE);
+        RenderContext renderContext = new RenderContext(null, null, node.getSession().getUser());
+        renderContext.setMainResource(resource);
+        renderContext.setServletPath("/cms/" + servlet);
+        JCRNodeWrapper nodeForURL = node;
+        if (findDisplayable) {
+            nodeForURL = JCRContentUtils.findDisplayableNode(node, renderContext);
+            if (nodeForURL != null && !nodeForURL.getIdentifier().equals(node.getIdentifier())) {
+                resource = new Resource(nodeForURL, "html", null, Resource.CONFIGURATION_PAGE);
+            }
+        }
+        Template template = RenderService.getInstance().resolveTemplate(resource, renderContext);
+        String extensionName;
+        if (template != null || nodeForURL.isNodeType("jnt:page") || nodeForURL.isNodeType("jmix:mainResource")) {
+            extensionName = ".html";
+        } else {
+            extensionName = ".content-template.html";
+        }
+
+        url += nodeForURL.getPath() + extensionName;
+
+        return url;
+    }
```

Comment:
Would add a TODO mentioning:

-  reference the original source code: [mention in comment that it's a copy/paste from Node](https://github.com/Jahia/jahia-private/blob/master/core/src/main/java/org/jahia/ajax/gwt/helper/NodeHelper.java#L102-L139)
- reference a tech ticket to move this code in a proper utility/service layer for later.